### PR TITLE
initial fuzzing harness

### PR DESCRIPTION
**Pull request by willscott**
* [x] allow dynamic specification of harnessed datastore implementation
* [x] include closing / re-opening of datastore in the script

### CODE REVIEW

**Comment by willscott** [comment]
@ribasushi if you have cycles, maybe you can take a first pass at review and see if this structure looks reasonable for a datastore fuzzing harness

File: fuzz/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+go run ./cmd/generate github.com/ipfs/go-ds-badger
+// or `go generate`
+DS_PROVIDERS="github.com/ipfs/go-ds-badger" go generate
+```
+
+Finally, build the fuzzing artifact, and fuzz
+```golang
+go-fuzz-build
+go-fuzz
+```
```
Comment (ribasushi) [inline fuzz/README.md:23]: Add a section at the very end:

If you have not installed the fuzzing tools yet, execute:
```
go get -u github.com/dvyukov/go-fuzz/go-fuzz github.com/dvyukov/go-fuzz/go-fuzz-build
```

File: fuzz/fuzzer.go [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+		dr := make(chan byte, 15)
+		go threadDriver(ctx, &state, dr)
+		state.inputChannels[i] = dr
+	}
+	return &state, nil
+}
+
+// Fuzz is a go-fuzzer compatible input point for replaying
+// data (interpreted as a script of commands)
+// to kown ipfs datastore implementations
```
Comment (ribasushi) [inline fuzz/fuzzer.go]:
```suggestion
// to known ipfs datastore implementations
```

File: go.mod [hunk 3]
```diff
@@ -1,10 +1,13 @@
 module github.com/ipfs/go-datastore
 
 require (
+	github.com/dvyukov/go-fuzz v0.0.0-20200318091601-be3528f3a813 // indirect
 	github.com/google/uuid v1.1.1
+	github.com/ipfs/go-ds-badger v0.2.1
+	github.com/ipfs/go-ds-leveldb v0.4.2
```
Comment (ribasushi) [inline go.mod]: do we want to add these as hard deps? Perhaps the fuzz portion needs to become a sub-project? @Stebalien thoughts?

**Review by ribasushi — APPROVED** [review]
This looks pretty excellently written! The instructions worked out of the box ( I made one suggestion to add, no edits )

I do not know the tooling too well, but from what I was able to read + running this locally: LGTM

File: go.mod (hunk 3 above)
Comment (Stebalien) [reply go.mod]: We don't. We should create a sub-module.

File: fuzz/go.mod [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module github.com/ipfs/go-datastore/fuzz
+
+go 1.14
+
+require (
+	github.com/ipfs/go-datastore v0.4.4
```
Comment (Stebalien) [inline fuzz/go.mod:8]: We can add a `replace` directive pointing to `../`.

File: fuzz/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+implementations.
+
+Usage
+----
+
+First, get the code
+```golang
+go get github.com/ipfs/go-datastore
+cd go-datastore/fuzz
+```
```
Comment (Stebalien) [inline fuzz/README.md]: This isn't going to work reliably (gopath, etc). I'd just remove it.

File: fuzz/cmd/compare/main.go [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	ds "github.com/ipfs/go-datastore"
+	fuzzer "github.com/ipfs/go-datastore/fuzz"
+	dsq "github.com/ipfs/go-datastore/query"
+
+	"github.com/spf13/pflag"
```
Comment (Stebalien) [inline fuzz/cmd/compare/main.go:12]: I'd slightly prefer to use github.com/urfave/cli to be consistent with the other projects.

**Review by Stebalien — APPROVED** [review]

File: fuzz/cmd/compare/main.go (hunk 6 above)
Comment (willscott) [reply fuzz/cmd/compare/main.go:12]: Given that i'm not pulling in a full app/cli framework for these entry points, i'm going to defer on that for now.

**Pull request merged**